aten/src/ATen/native/hammerblade/Cat.cpp

```diff
@@ -0,0 +1,193 @@
+#include <ATen/ATen.h>
+#include <ATen/hammerblade/HammerBladeContext.h>
+#include <ATen/native/hammerblade/Offload.h>
+#include <ATen/native/Copy.h>
+#include <ATen/native/TensorIterator.h>
+#include <ATen/native/hammerblade/Resize.h>
+#include <ATen/native/ResizeCommon.h>
+
+namespace at { namespace native {
+
+Tensor& resize_h_(
+    Tensor& self,
+    IntArrayRef size,
+    c10::optional<MemoryFormat> optional_memory_format) {
+#ifdef BUILD_NAMEDTENSOR
+  if (self.has_names()) {
+    return resize_named_tensor_(self, size, optional_memory_format);
+  }
+#endif
+  auto* self_ = self.unsafeGetTensorImpl();
+  resize_impl_hb_(self_, size, /*strides=*/c10::nullopt);
+  self_->maybe_zero_dim(size.size() == 0);
+  if (optional_memory_format.has_value()) {
+    auto memory_format =
+        optional_memory_format.value();
+    TORCH_CHECK(
+        memory_format != MemoryFormat::Preserve,
+        "Unsupported memory format",
+        memory_format);
+    self_->empty_tensor_restride(memory_format);
+  }
+  return self;
+}
+
+Tensor _cat_hb(TensorList tensors, int64_t dim) {
+
+  std::cout << "in _cat_hb, TensorList size = " << tensors.size()
+            << " dim = " << dim << std::endl;
+  // convert TensorList length to uint32
+  uint32_t length_u32 = safe_downcast<uint32_t, size_t>(tensors.size());
+  // convert dim to int32
+  int32_t dim_i32 = safe_downcast<int32_t, int64_t>(dim);
+
+  // plain tensors
+  std::vector<Tensor> tensor_args;
+
+  // scalars
+  std::vector<eva_t> scalars;
+  scalars.push_back(create_device_scalar(length_u32));
+  scalars.push_back(create_device_scalar(dim_i32));
+
+  // XXX: this is definitely wrong
+  Tensor result = at::empty({0}, tensors[0].options());
+  bool allSkipped = true;
+  bool allContiguous = true;
+  Tensor notSkippedTensor;
+
+  // Inputs cannot alias the output tensor
+  for (size_t i = 0; i < tensors.size(); i++) {
```

Comment:
This piece of code looks strange. `result` is just allocated above `Tensor result = at::empty({0}, tensors[0].options());`, so it is not possible for it to alias with anybody.

Reply:
updated Cat.cpp. I implemented host code based on functions in TensorShape.cpp